src/decoder/decoder.ml

```diff
@@ -247,37 +247,36 @@ let can_decode_kind decoded_type target_kind =
            true
          with Unsupported -> false)
   in
-  match Frame.type_of_kind target_kind with
+  match target_type with
     (* Either we can decode straight away. *)
-    | _ when Frame.type_has_kind decoded_type target_kind -> true
+    | _ when decoded_type = target_type -> true
     (* Or we can convert audio and/or drop video and midi *)
     | { Frame.audio; video; midi } ->
         let audio =
           if can_convert_audio audio then audio else decoded_type.Frame.audio
         in
         let video = if video = 0 then 0 else decoded_type.Frame.video in
         let midi = if midi = 0 then 0 else decoded_type.Frame.midi in
-        Frame.type_has_kind { Frame.audio; video; midi } target_kind
-
-let decoder_modes kind =
-  Frame.(
-    match (kind.audio, kind.video, kind.midi) with
-      | Succ _, Succ _, Zero -> [`Audio_video]
-      | Succ _, _, Zero -> [`Audio; `Audio_video]
-      | _, Succ _, Zero -> [`Video; `Audio_video]
-      (* Now entering the annoying part. *)
-      | Any, Any, Zero -> [`Audio_video]
-      | Any, Zero, Zero -> [`Audio; `Audio_video]
-      | Zero, Any, Zero -> [`Video; `Audio_video]
-      | Zero, Zero, Zero -> []
-      | Zero, Zero, _ -> [`Midi]
-      | _ -> [])
+        target_type = { Frame.audio; video; midi }
+
+let decoder_modes ctype =
+  let audio = ctype.Frame.audio in
+  let video = ctype.Frame.video in
+  let midi = ctype.Frame.midi in
+  if audio = 0 && video = 0 && midi = 0 then
+    [`Audio_video; `Audio; `Video; `Midi]
```

Comment:
```suggestion
    [`Metadata]
```

Let's make a `` `Metadata`` decoder type for this case.

Reply:
I used another way to solve the thing (`file.metadata` directly requests resolution of metadata), but you should be pleased: we now have `[]` as answer for this case.